### ISSUE TITLE
Multi-Dimensional Arrays?

### ISSUE THREAD

**Issue by mmstick** [redox-os/ion#470]
As I already have plans laid out to implement multi-dimensional associative arrays in the future, we could also use similar methods for creation of multi-dimensional arrays. It's basically just enabling the ability to parse elements as if they were arrays themselves.

```ion
$ let md_array = [ "[one two three]" "[four five six]" "[seven eight nine]"]
$ echo md_array[0]
> [one two three]
$ echo md_array[0][0]
> one
```

**Comment by huntergoldstein** [comment]
Why do the elements need to be quoted?

**Comment by mmstick** [comment]
If they aren't quoted, they will be flattened.  They are quoted to store as strings, and if a second index is called for, that string will be evaluated to see if it contains an array as well.

**Comment by mmstick** [comment]
I guess we could add more sophisticated parsing to the `let` builtin so that the quotes aren't a hard requirement.

**Comment by bb010g** [comment]
What about a different array type for storing boxed expressions (strings, arrays, hashmaps, &c.)? Something like `@[ [1 2 3] foo @[ 4 [5] 6]]` could work. Not having to parse elements repeatedly would be good for performance.

**Comment by mmstick** [comment]
If we had a `@[]` though, what would `$[]` be?

**Comment by bb010g** [comment]
There wouldn't be one, which is why I'm not too set on that sigil for the syntax. It's just the simplest thing I could think of for an example. I also don't like how similar it looks to command substitution.

**Comment by mmstick** [comment]
Yeah, I think I'll reserve `@[]` and `$[]` for a different feature later down the line. Perhaps for use by plugins. Here's one idea that I have:

### Optionally-Typed Identifier

This could go hand in hand with the optionally-typed identifiers that I have planned for the `let` command so that it will be possible to set multiple variables and arrays in the same statement. We could use an identifier to specify that we want to parse a multi-dimensional array, and then therefore search for that pattern (returning an error if the provided string isn't a valid pattern).

We could simply search for, at minimum, a `[[` pattern that is ended with `]]` for a 2D array, or `[[[` that is ended with `]]]` for a 3D array, and so on.

```ion
let multi_array[[]] = [[ one two ] [ three four ] [ five six ]]
echo @multi_array[1][0]
```

**Comment by mmstick** [comment]
This is now a possible implementation path now that I've landed the new assignment logic.

**Comment by covercash2** [comment]
i have a implementation that needs a few more tests in [this commit](https://github.com/covercash2/ion/commit/defc064a10ebcddc1aea10d8b6e86ba2ebcaf456).

i had a question about expanding multi-dimensional arrays. my naive implementation expanded like this:

    $ let a = [1 [2 3] [4 5] 6]
    $ echo @a 
    1 [ 2 3 ] [ 4 5 ] 6

that implementation doesn't pass the .ion script tests. those tests expect multi-dimensional `a` to expand to `1 2 3 4 5 6`.

do we want to flatten arrays this way every time? maybe include a method to specify the type of output. i'm working on flattening the arrays (pretty easy, it'll be done shortly), but i just wanted to get confirmation.

**Comment by mmstick** [comment]
Flattening by default likely makes the most sense.

**Comment by covercash2** [comment]
i'm a little stuck here.

the way i've designed the Element is to be built from an expanded String, str, or string Array. but since `expand_string` is called recursively and only returns a flat array, there has to be some way to parse an Element from that flat array, which there is, but it requires the square brackets for the parser to know how to parse the sub-array elements. also, `fn expand_string` is a huge choke point; it's used everywhere. i can't just change the return type of expand_string so easily.

now if Elements are _not_ flattened by default, then the current behavior of most things would remain the same. it doesn't require any extra dependencies in `parser::shell_expand`. it allows for arrays to be passed as strings within the context of scripting, etc., and be parsed back into Elements. in addition, i can rewrite only a few tests to reflect this behavior, and show green across the board, unit and integration. 

i dunno. something to consider. let me know. i'll keep stewing on it.
